src/pb_tensor.cc

```diff
@@ -231,6 +232,52 @@ PbTensor::FromNumpy(const std::string& name, py::array& numpy_array)
   return std::make_shared<PbTensor>(name, numpy_array);
 }
 
+DLDeviceType
+PbTensor::DeviceType()
+{
+  DLDeviceType device_type{};
+
+  switch (memory_type_) {
+    case TRITONSERVER_MEMORY_GPU:
+      device_type = DLDeviceType::kDLCUDA;
+      break;
+    case TRITONSERVER_MEMORY_CPU:
+      device_type = DLDeviceType::kDLCPU;
+      break;
+    case TRITONSERVER_MEMORY_CPU_PINNED:
+      device_type = DLDeviceType::kDLCUDAHost;
+      break;
+  }
+
+  return device_type;
+}
+
+py::capsule
+PbTensor::DLPack(const py::object& stream)
+{
+// Array API requirements for the stream argument:
+// stream = None, producer must assume the legacy default stream,
+// stream = -1 is a signal for the producer not to perform any synchronization
+// stream = 1 the legacy default stream (in this case should synchronize
+// on CUDA stream 0)
+#ifdef TRITON_ENABLE_GPU
+  if (!IsCPU() && !stream.is(py::int_(-1))) {
+    cudaStream_t cuda_stream =
+        (stream.is(py::int_(1)) || stream.is(py::none()))
+            ? 0
+            : reinterpret_cast<cudaStream_t>(py::cast<uint64_t>(stream));
+    cudaError_t err = cudaStreamSynchronize(cuda_stream);
```

Comment:
We shouldn't synchronize their stream. I think to guarantee stream ordering we don't have to do anything with their stream since the pointers are already synchronized.

Reply:
@Tabrizian  I don't think that pointers are already synchronized at this point. `DLPack` is invoked by external tensor's `__dlpack__` method. We want to make sure that pointer contains correct data and no pending job is left, that can corrupt this data, right? if not `cudaStreamSunchronize`, we should at least create an event on their stream and wait for this event on our stream.

Reply:
I don't think that we can assume that pointers are synchronized at this point. When external stream calls pbTensor's `__dlpack__` method, which invokes this function, I believe we are not guaranteed that pointers will be sync'ed.  In older DLPack implementations with intermediate capsule, I believe some implementations assumed that `to_dlpack` and `from_dlpack` are on the same stream, thus they are aligned. in here, `__dlpack__` is called on the external tensor's stream and is accessing `pbTensor` on the legacy stream, so we must ensure that any pending operations on the default stream are finished. I agree that  current implementation is not the right way of ensuring that, but we still need to make sure that all work on `pbTensor` is done, before passing it to external tensor